### DISCUSSION TITLE
NextJS Best Practices example with Client Extensions

### DISCUSSION THREAD

**Discussion by kroucher** [prisma/prisma#20321] [Q&A] [answered]
Hoping for a better understanding of how to use Client Extensions using the instantiated client from the [NextJS Best Practices guide](https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices).

Currently, the type of the exported client is being overridden with the following:
```ts
const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined
}
```
so when the client is extended with
```ts
export const prisma = globalForPrisma.prisma ?? new PrismaClient().$extends({
  result: {
    user: {
      fullName: {
        needs: { firstName: true, lastName: true },
        compute(user) {
          return `${user.firstName} ${user.lastName}`
        },
      },
    },
  },
})
```
Due to the exported variable being overridden when not in `production` (with `.. as { prisma: PrismaClient | undefined }`), Typescript still thinks `prisma` is just `PrismaClient`, not the extended `DynamicClientExtensionThis<Prisma.TypeMap<Args & {..` and has no knowledge of the new field.

If the Client Extension is added like this, Typescript is happy, but does not include the prevention of multiple clients being instantiated:

```ts
const prismaClient =
  globalForPrisma.prisma ||
  new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });

const prisma = prismaClient.$extends({
  result: {
    computerConfig: {
      item: {
        compute() {
          return "Hello world!";
        },
      },
    },
  },
});

export { prisma };
```

**Comment by ludralph** [comment]
Hi @kroucher 👋 

Thank you for raising this issue.

We have tried reproducing this error with the Prisma examples [Nextjs repository](https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes) but did not get the exact same error you mentioned. We tried instantiating the Prisma client like this:
```ts
// prisma.ts 

const globalForPrisma = global as unknown as { prisma: PrismaClient }

export const prisma = globalForPrisma.prisma || new PrismaClient().$extends({
    result: {
      user: {
        fullName: {
          needs: { name: true, email: true },
          compute(user) {
            return `${user.name} ${user.email}`
          },
        },
      },
    },
  })

if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma
```
We have then used the new `fullName` field like this:
```tsx
// index.tsx
export const getServerSideProps: GetServerSideProps = async () => {
  const feed = await prisma.post.findMany({
    where: { published: true },
    select:{
      id: true,
      title: true,
      content: true,
      published: true,
      author:{
        select:{
          id: true,
          fullName: true
        }
      }
    }
  })
  console.log(feed);
  
  return {
    props: { feed },
  }
}
```
With this query, we saw the following error inside of VS Code:
```
Type '{ id: true; fullName: true; }' is not assignable to type 'UserSelect<DefaultArgs>'.
  Object literal may only specify known properties, and 'fullName' does not exist in type 'UserSelect<DefaultArgs>'
```
Was this the same error that you encountered? We were still able to run the code. The typescript error did not block us from executing it.

It would be helpful if you could provide a reproduction so that we can look further into the issue.

**Reply by kroucher** [reply]
Hi @ludralph, thanks heaps for the fast response. 

> With this query, we saw the following error inside of VS Code:

Correct, this is the Typescript compiler letting us know that according to the types it's been given, it has no knowledge of `fullName` on the `User` model. This is because of the instantiation of the global variable `prisma` with `const globalForPrisma = global as unknown as { prisma: PrismaClient }`. From my understanding, this typecasting tells the Typescript compiler that if anything refers to `prisma`, the type definitions for `prisma` is (an unaltered, unextended) `PrismaClient`. After this variable instantiation, we proceed to modify what `prisma` actually is.

To see this in practice, using the example repo linked above, if you comment out the instantiation of the global variable, and both assignments of the global variable, hovering over the `prisma` variable shows a very different type than `PrismaClient`:

Before commenting:
 
![image](https://github.com/prisma/prisma/assets/99768034/930ccde2-7c94-47ff-8f61-c5d43b38e920)

Hover before commenting:

![image](https://github.com/prisma/prisma/assets/99768034/d83942dc-b4f6-4017-96dd-9778fce9a8e1)

After commenting: 

![image](https://github.com/prisma/prisma/assets/99768034/8e0a0671-51b7-4d2b-b1ea-4d86eaaf4fff)

Hover after commenting: 

![image](https://github.com/prisma/prisma/assets/99768034/a4f35a5d-6b12-44a8-b50d-3f15416dd839)

> Was this the same error that you encountered?

Not quite the same error but due to the same cause. In the repo you linked above, in the same `getServerSideProps`, adding the following lines will also highlight the issue:

```ts
  const user = await prisma.user.findUnique({
    where: { id: 1 },
  })

  const userFullName = user?.fullName;
                              //^ Property 'fullName' does not exist on type 'GetResult<{ id: number; email: string; name: string | null; }, unknown> & {}'
  ```

However - by making the changes I mentioned above in regards to commenting out the globally instantiated `prisma` variable, and let Typescript inference take over, all of the errors mentioned here go away, with the code working as expected:

//index.tsx

Before commenting out:

![image](https://github.com/prisma/prisma/assets/99768034/3e9916b6-2a03-4043-b651-390a40d9fa6f)

After commenting out:

![image](https://github.com/prisma/prisma/assets/99768034/d63241f0-3acc-41f3-82a3-acd676173471)

With these examples, we can see that by setting the type of `prisma` to be `PrismaClient` as part of the global instantiation of the client results in Typescript being unaware of the changes/client extensions being added, and thus altering the type that is being consumed elsewhere. However removing/commenting out the global instantiation is not ideal either, as its purpose is due to the `There are already 10 instances of Prisma Client..` error discussed in #5103.

Ideally it would be best to come up with a solution that avoids the global instantiation, as it seems to be more of a bandaid to the problem and is now causing Typescript errors because we are typecasting the global variable to one type `PrismaClient` before altering it to another `DynamicClientExtensionThis<Prisma.TypeMap<Args & {...`. 

Thanks again, and let me know if there is any further information I can provide.

**Comment by ovidyou** [comment, marked as answer]
@kroucher did you find a solution? I think the code below should work for your use case:

```typescript
import { PrismaClient } from "@prisma/client";

const getExtendedClient = () => {
  return new PrismaClient({
    log: process.env.NODE_ENV === "development" ? ["query", "error", "warn"] : ["error"],
  }).$extends({
    result: {
      user: {
        fullName: {
          needs: { firstName: true, lastName: true },
          compute(user) {
            return `${user.firstName} ${user.lastName}`;
          },
        },
      },
    },
  });
};

type ExtendedPrismaClient = ReturnType<typeof getExtendedClient>;

// PrismaClient is attached to the `global` object in development to prevent
// exhausting your database connection limit.
//
// Learn more:
// https://pris.ly/d/help/next-js-best-practices
const globalForPrisma = globalThis as unknown as {
  prisma: ExtendedPrismaClient | undefined;
};

export const prisma = globalForPrisma.prisma ?? getExtendedClient();

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;

```

**Reply by kroucher** [reply]
Hi @ovidyou, thanks so much for the response - sorry for the delay getting back on this, I've tested this solution and it works a charm 🥳

Thanks a bunch for your time 🤝
